scheduler/pkg/kafka/dataflow/server_test.go

```diff
@@ -475,6 +475,165 @@ func TestPipelineEvents(t *testing.T) {
+		status         pipeline.PipelineStatus
+		expectedStatus pipeline.PipelineStatus
+		expectedOp     chainer.PipelineUpdateMessage_PipelineOperation
+		connection     bool
+	}
+
+	tests := []test{
+		{
+			name: "rebalance ready pipeline",
+			loadReq: &scheduler.Pipeline{
+
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*scheduler.PipelineStep{
+					{
+						Name: "a",
+					},
+				},
+			},
+			status:         pipeline.PipelineReady,
+			expectedStatus: pipeline.PipelineCreating,
+			expectedOp:     chainer.PipelineUpdateMessage_Create,
+			connection:     true,
+		},
+		{
+			name: "rebalance creating pipeline",
+			loadReq: &scheduler.Pipeline{
+
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*scheduler.PipelineStep{
+					{
+						Name: "a",
+					},
+				},
+			},
+			status:         pipeline.PipelineCreating,
+			expectedStatus: pipeline.PipelineCreating,
+			expectedOp:     chainer.PipelineUpdateMessage_Create,
+			connection:     true,
+		},
+		{
+			name: "rebalance terminating pipeline",
+			loadReq: &scheduler.Pipeline{
+
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*scheduler.PipelineStep{
+					{
+						Name: "a",
+					},
+				},
+			},
+			status:         pipeline.PipelineTerminating,
+			expectedStatus: pipeline.PipelineTerminating,
+			expectedOp:     chainer.PipelineUpdateMessage_Delete,
+			connection:     true,
+		},
+		{
+			name: "rebalance terminating pipeline - no connection",
+			loadReq: &scheduler.Pipeline{
+
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*scheduler.PipelineStep{
+					{
+						Name: "a",
+					},
+				},
+			},
+			status:         pipeline.PipelineTerminating,
+			expectedStatus: pipeline.PipelineTerminated,
+			connection:     false,
+		},
+		{
+			name: "no connection",
+			loadReq: &scheduler.Pipeline{
+
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*scheduler.PipelineStep{
+					{
+						Name: "a",
+					},
+				},
+			},
+			status:         pipeline.PipelineReady,
+			expectedStatus: pipeline.PipelineCreate,
+			connection:     false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			serverName := "dummy"
+			s, _ := createTestScheduler(t, serverName)
+			stream := newStubServerStatusServer(20)
+			if test.connection {
+				s.streams[serverName] = &ChainerSubscription{
+					name:   "dummy",
+					stream: stream,
+					fin:    make(chan bool),
+				}
+				g.Expect(s.streams[serverName]).ToNot(BeNil())
+			}
+
+			err := s.pipelineHandler.AddPipeline(test.loadReq) // version 1
+			g.Expect(err).To(BeNil())
+
+			//to allow events to propagate
+			time.Sleep(500 * time.Millisecond)
+
+			err = s.pipelineHandler.SetPipelineState(test.loadReq.Name, test.loadReq.Version, test.loadReq.Uid, test.status, "", sourceChainerServer)
+			g.Expect(err).To(BeNil())
+
+			//to allow events to propagate
+			time.Sleep(500 * time.Millisecond)
```

Comment:
Are these the ugly sleeps you were telling me about? Not sure they are avoidable given the nature of the local (async) event hub. But (note to self) we should probably do some concurrent operation tests on a single object to make sure the state remains consistent. I believe we could even do that under the current consistency checking k6 test, by having max 1 model/pipeline.